### DISCUSSION TITLE
Deploying to Vercel in standalone mode results in 'Code: NOT_FOUND'

### DISCUSSION THREAD

**Discussion by jschuur** [keystonejs/keystone#8254] [Questions] [answered]
Should I be able to deploy just the admin UI/GraphQL endpoint to Vercel so that both can be run remotely? I know there are instructions on doing an [embedded deployment](https://keystonejs.com/docs/walkthroughs/embedded-mode-with-sqlite-nextjs#bonus-deploy-to-vercel), but this doesn't allow you to use the admin UI online, since that runs on another port.

I ran `npx create-keystone-app` and then linked a repo for that to a Vercel app, picked `keystone build` as the build command and `.keystone/admin/.next` as the output dir, but after it deploys, I get an [error](https://keystone-test-hl6pe3tib-joostschuur.vercel.app/) when visiting the Vercel URL (same if I use `.keystone/admin/` for the output dir):

```
404: NOT_FOUND
Code: NOT_FOUND
```

If I look at the 'Source / Output' tab for the deployment on Vercel, the directory structure of what I see looks very different, so I am guessing it's not supposed to be this easy.

**Comment by mfxart** [comment]
@jschuur any updates on it? Having the same problem right now

**Comment by dkhierl-chacrasoftware** [comment]
+1

**Comment by jschuur** [comment, marked as answer]
When I originally posted this as an issue, I misunderstood that the CMS part wasn't necessarily designed to be deployed to Vercel directly.

There's [an example](https://github.com/keystonejs/keystone/tree/main/examples/nextjs-keystone) on how to get the GraphQL API wrapped up in a Next.js frontend app, but you don't get the Admin UI as part of the same instance.

It [doesn't sound like](https://github.com/keystonejs/keystone/tree/main/examples/nextjs-keystone#2-what-should-i-do-to-both-use-keystone-in-my-nextjs-app-and-have-a-fully-functioning-admin-ui) you can get the Keystone server/admin UI to run on Vercel, so for now, I just started paying Heroku to continue to host it there until I can spend time trying to run this for free on Railway e.g..

**Reply by flexdinesh** [reply]
Keystone is an extended express server. [Vercel recommends against using standalone express servers](https://vercel.com/guides/using-express-with-vercel) in their infra because of **standalone server vs serverless** architecture tradeoffs. 

The [nextjs-keystone example](https://github.com/keystonejs/keystone/tree/main/examples/nextjs-keystone) uses composable parts of Keystone (GraphQL layer, data APIs, etc) within a Next.js app without starting an express server and that's why we could easily deploy it to Vercel. But our Admin UI is not yet composable with an external Next.js app, hence the limitation - can't use Admin UI in the example in Vercel. 

Our current recommendation is that - if you want to be able to use both Admin UI and GraphQL API, deploy to server instances like Heroku, DO, etc. If you want to be able to deploy to serverless infra like Vercel, there might be tradeoffs and you might have to do some customisations on your own because there's a very good chance that the code won't work OOTB. 

If anyone successfully gets Keystone to work in serverless infra without tradeoffs, we're keen to hear about it. In the meanwhile, we have gotten our data APIs to work in serverless infra, next step is to get Admin UI to work in serverless infra. The team is aware of this tradeoff and are already thinking in the direction to make it easy to use both the GraphQL API and Admin UI in serverless infra.

**Reply by TheMikeyRoss** [reply]
The example link seems to be broken

**Comment by phifa** [comment]
this would be so nice to have! @flexdinesh  any news?
